### ISSUE TITLE
Can't break on any client-side breakpoint with Debugger for Chrome

### ISSUE THREAD

**Issue by TitaneBoy** [Microsoft/vscode-chrome-debug#313]
Versions:
Debugger for Chrome: 2.4.1
Visual Studio Code: 1.8.0
Node.js: 6.4.0
AngularJS: 1.6.0
Express: 4.14.0
OS: Windows 7 Pro 64 bits.

Hi. I'm facing issue when trying to use the Debugger for Chrome.
I'm trying to use it as simple as possible, with minimum requierement.

I've built an nodejs application which has a server (with express) and client side. In client side, I'm using angularJS. The whole thing works very well. I can see my web pages as expected. Both client side and server side work well.

But for some reason, I can't debug client side application.

Here is the configuration I'm using for the client side:

``` 
   {
        "name": "Launch client web with sourcemaps",
        "type": "chrome",
        "request": "launch",
        "url": "http://localhost",
        "webRoot": "${workspaceRoot}",
        "sourceMaps": true
    }
```

I've separated my controllers and factory in single files in order to be consistent. And as I said before, every thing looks to work fine because I can see what is expected. But when I put a breakpoint in a controller function for example, I can't break on it whatever I'm doing to get inside this function.

Can you help me please and tell me what I'm doing wrong ?

Thank you very much for your help.

PS: I've killed all chrome instance in the task manager and I added " --remote-debugging-port=9222" in the chrome target before to launch the debugger.

**Comment by TitaneBoy** [comment]
I noticed that when the debugger is launched, my breakpoints are disactived and when I put the mouse over the breakpoint, It said "Breakpoint is ignored because generated code not found (source map problem ?.)".

Does it means that I need to generate myself source map files for all my javascript files ?

**Comment by TitaneBoy** [comment]
Here is the diagnostic logging I have
[vscode-chrome-debug.txt](https://github.com/Microsoft/vscode-chrome-debug/files/661688/vscode-chrome-debug.txt)

**Comment by roblourens** [comment]
It appears that your scripts don't have sourcemaps - your build process needs to generate those too.

**Comment by TitaneBoy** [comment]
Hi. Thank you for your answer.
Is there a standard way to generate sourcemaps files from our scripts ?  I've tried to generate a map with this:  [Closure Compiler
](https://developers.google.com/closure/compiler/docs/gettingstarted_app), but even I had map file, I could not break inside my javascript file. 

Once I have map files, what I have to do next ? I've put them at the same level than original javascript file and tryied again to break inside source file but it didn't worked for me. I've tried to find a solution on the web but I could not make it work.  Could you help me please ?

Thank you for your answer

**Comment by roblourens** [comment]
The sourcemap section of the readme may be helpful - also try using the '.scripts' command to print out the debugger's info on your sources.

**Comment by TitaneBoy** [comment]
Hi @roblourens. I've given another try to be able to debug client side javascript files.
In my current case, I would like to test it only on 1 js file. It is located to ${workspaceRoot}/scripts/factory/posts.js

I've generated the map file for this file (posts.js) and when looking at the diagnostic logging file (and the .script command), it look that my map file is founded and used...but somehow, I can't break on any breakpoint in my posts.js file in visual studio code.

Here is my diagnostic file:
[vscode-chrome-debug.txt](https://github.com/Microsoft/vscode-chrome-debug/files/702681/vscode-chrome-debug.txt)

And here is the output of the .script command:

```
› eval://84
› eval://99
› http://localhost/
› http://localhost/ang_script.js
› http://localhost/angular-route/angular-route.min.js
    - angular-route.js (c:\Personal_Data\Cortex_Setup\Node.js\Express\Angular_JS_Demo\angular-route\angular-route.js)
› http://localhost/angular.min.js
    - angular.js (c:\Personal_Data\Cortex_Setup\Node.js\Express\Angular_JS_Demo\angular.js)
› http://localhost/controller/comments.js
› http://localhost/controller/posts.js
› http://localhost/directive/comments.js
› http://localhost/directive/datepicker.js
› http://localhost/directive/tabs.js
› http://localhost/directive/time.js
› http://localhost/factory/posts.js
    - posts.js (c:\Personal_Data\Cortex_Setup\Node.js\Express\Angular_JS_Demo\factory\posts.js)
› http://localhost/jquery-3.1.1.min.js
› http://localhost/picker.date.js
› http://localhost/picker.js
```

And As you can see, the "factory\posts.js" file looks to be found.

May be I'm missing something else to make it works...Could you help me please with that ?

Here's my configuration:
```
{
        "name": "Launch client web with sourcemaps",
        "type": "chrome",
        "request": "launch",
        "url": "http://localhost",
        "webRoot": "${workspaceRoot}",
        "diagnosticLogging":true,
        "sourceMaps": true
    }
```

Here is my map file for the factory/posts.js file: 
[posts.js.map.txt](https://github.com/Microsoft/vscode-chrome-debug/files/702706/posts.js.map.txt)

Thank you again for your answer and your help

**Comment by roblourens** [comment]
Is `c:\Personal_Data\Cortex_Setup\Node.js\Express\Angular_JS_Demo\factory\posts.js` the correct path of the source file? You said it was under the 'scripts' directory but I don't know if that's the source or the output file. 

Is the webserver root in workspaceRoot, or workspaceRoot/scripts?

And your sourcemap only says that the source is located at `posts.js` which is probably not enough info to find it.

**Comment by TitaneBoy** [comment]
Well..You're wright...The correct path of the source file should be `c:\Personal_Data\Cortex_Setup\Node.js\Express\Angular_JS_Demo\scripts\factory\posts.js`

I don't understand why it is saying `c:\Personal_Data\Cortex_Setup\Node.js\Express\Angular_JS_Demo\factory\posts.js`

There is no folder named "factory" at the `c:\Personal_Data\Cortex_Setup\Node.js\Express\Angular_JS_Demo\` directory. The factory folder is, as I said, at `c:\Personal_Data\Cortex_Setup\Node.js\Express\Angular_JS_Demo\scripts\`.

 So why he is pointing at the wrong path ?

**Comment by roblourens** [comment]
Because your webRoot is workspace, and the url is `http://localhost/factory/posts.js`, and the sourcemap only says that the relative path to the source file is 'posts.js'. So the conclusion from combining these pieces of info is that path.

**Comment by TitaneBoy** [comment]
Hi @roblourens . Thank your for your answer. You were right. I've modified the webRoot value for "${workspaceRoot}/scripts" and it looks better.

But I thing I still have an issue.

Here my new diagnostic file:  
[vscode-chrome-debug.txt](https://github.com/Microsoft/vscode-chrome-debug/files/704461/vscode-chrome-debug.txt)

At the lines 51-53, we have:

```
Paths.scriptParsed: resolved http://localhost/factory/posts.js to c:\Personal_Data\Cortex_Setup\Node.js\Express\Angular_JS_Demo\scripts\factory\posts.js. webRoot: c:\Personal_Data\Cortex_Setup\Node.js\Express\Angular_JS_Demo/scripts
SourceMaps.getMapForGeneratedPath: Finding SourceMap for c:\Personal_Data\Cortex_Setup\Node.js\Express\Angular_JS_Demo\scripts\factory\posts.js by URI: posts.js.map and webRoot: c:\Personal_Data\Cortex_Setup\Node.js\Express\Angular_JS_Demo/scripts
SourceMaps.loadSourceMapContents: Reading local sourcemap file from c:\Personal_Data\Cortex_Setup\Node.js\Express\Angular_JS_Demo\scripts\factory\posts.js.map
```

And at the line 80, we can see this line:
`SourceMaps.scriptParsed: c:\Personal_Data\Cortex_Setup\Node.js\Express\Angular_JS_Demo\scripts\factory\posts.js was just loaded and has mapped sources: ["c:\\Personal_Data\\Cortex_Setup\\Node.js\\Express\\Angular_JS_Demo\\scripts\\factory\\posts.js"]`

So I guess we can say that the map source was found for the posts.js file.

Starting from line 113, we can see informations about breakpoints set to the posts.js file. But at line 125, we can see this:

`SourceMaps.setBP: Can't map c:\Personal_Data\Cortex_Setup\Node.js\Express\Angular_JS_Demo\scripts\factory\posts.js:3:19, keeping the line number as-is.
SourceMaps.setBP: Can't map c:\Personal_Data\Cortex_Setup\Node.js\Express\Angular_JS_Demo\scripts\factory\posts.js:7:9, keeping the line number as-is.
SourceMaps.setBP: Can't map c:\Personal_Data\Cortex_Setup\Node.js\Express\Angular_JS_Demo\scripts\factory\posts.js:7:9, keeping the line number as-is.`

I still can't break on the posts.js file. And I don't know if you have an idea of why.

Here is my thoughts:
I used Google [Closure Compiler Application](https://developers.google.com/closure/compiler/docs/gettingstarted_app) to generate map file from javascript file. I use this command:

`java -jar compiler.jar \ 
     --js script.js \
     --create_source_map ./script-min.js.map \
     --source_map_format=V3 \
     --js_output_file script-min.js
` as described here [source](https://www.html5rocks.com/en/tutorials/developertools/sourcemaps/).

If I understand well, it minify the source (posts.js) first and then generate a map file.

My posts.js file has 67 lines. You can see it here: 
[posts.js.txt](https://github.com/Microsoft/vscode-chrome-debug/files/704529/posts.js.txt)

The generated minified file is here: 
[posts-min.js.txt](https://github.com/Microsoft/vscode-chrome-debug/files/704531/posts-min.js.txt)

And the generated map file is here: 
[posts.js.map.txt](https://github.com/Microsoft/vscode-chrome-debug/files/704533/posts.js.map.txt)

I don't know if I understand well, but by looking in the map file, it looks that this file is based on the minified file generated by the Closure Compiler. So the map file is a map of the posts-min.js file, not the posts.js file. But when I put breakpoints on the posts.js file, it could not make a link between the map and the source. I'm saying that because the mappings value in the map file looks to be based on only 2 lines (which is the number of lines of the minified file)

Does what I say make sense ? If yes, what is the best way to generate map file based on javascript file (not minified) ? The [Debugger for Chrome SourceMaps section](https://github.com/Microsoft/vscode-chrome-debug#sourcemaps) doesn't say how can we generate map file unfortunetly and the only way I found to generate map file is with the [Closure Compiler application](https://developers.google.com/closure/compiler/) , from Google.

Thank you again for your answer, help and advices

**Comment by roblourens** [comment]
There are lots of different build processes out there, and I don't know anything about the closure compiler, so I can't help you with generating a sourcemap. But I don't understand your setup, are you not using the minified file on the page?

By the way, if sourcemap debugging with your setup doesn't work with chrome devtools, it won't work here.

**Comment by TitaneBoy** [comment]
On my page, I don't use the minified file. I use the originale source file posts.js.txt (without the .txt extension of course) posted previously.
It looks that chrome devtools works fine without my genrated source map file. If I use it, it just breaks at the first line of the original source file (posts.js). I think there is something wrong with the map file. I still don't understand how to generate them properly :-(

**Comment by TitaneBoy** [comment]
I finally found a way to make it works... I used npm package [uglify-js](https://www.npmjs.com/package/uglify-js) and generate the map file with this command:

`uglifyjs posts.js -o posts-min.js --source-map posts.js.map --source-map-root /factory/`

I saw somewhere on internet that we must modify the source file by adding at the end this line:
`//# sourceMappingURL=...` 

But at all, no need to add this line at the end of the original source file. So with all that, I could make it work great. I guess I also needed before, in my map file, to have the tag sourceRoot (added by the uglify option --source-map-root) to make it work. 

Well...Thank you again for your help. It was hard to figure out how all of this works, but at the end, everything works fine. Thank you.
